File: ametras_fast_import_addon/vue-app/src/importer/idempotency.ts
import type { ParsedRow } from './csvParser'

export type TimeoutIdempotencyKeyType = 'id' | '.id' | null

export interface TimeoutRetryIdempotencyAssessment {
  keyType: TimeoutIdempotencyKeyType
  keyColumn: string | null
  safeRows: ParsedRow[]
  unsafeRows: ParsedRow[]
  unsafeReasonCounts: Record<string, number>
}

function findMappedCsvColumn(
  fieldMappings: Record<string, string>,
  targetField: 'id' | '.id'
): string | null {
  for (const [csvCol, odooField] of Object.entries(fieldMappings)) {
    if (odooField === targetField) return csvCol
  }
  return null
}

function countReason(map: Record<string, number>, key: string): void {
  map[key] = (map[key] || 0) + 1
}

function readTrimmedValue(row: ParsedRow, column: string | null): string {
  if (!column) return ''
  const raw = row.data[column]
  return typeof raw === 'string' ? raw.trim() : ''
}

function isValidDatabaseId(value: string): boolean {
  return /^\d+$/.test(value) && Number(value) > 0
}

/**
 * Row-level timeout idempotency assessment for standalone mode.
 *
 * Mapping-level checks are not sufficient: retries are safe only when each row
 * actually carries a usable key value (id or .id). When both are mapped, rows
 * may safely retry with either key:
 * - prefer non-empty external ID ('id')
 * - otherwise fallback to valid database ID ('.id')
 */
export function assessTimeoutRetryIdempotency(
  rows: ParsedRow[],
  fieldMappings: Record<string, string>
): TimeoutRetryIdempotencyAssessment {
  const externalIdColumn = findMappedCsvColumn(fieldMappings, 'id')
  const dbIdColumn = findMappedCsvColumn(fieldMappings, '.id')

  const keyType: TimeoutIdempotencyKeyType =
    externalIdColumn ? 'id' : dbIdColumn ? '.id' : null
  const keyColumn = externalIdColumn || dbIdColumn

  const safeRows: ParsedRow[] = []
  const unsafeRows: ParsedRow[] = []
  const unsafeReasonCounts: Record<string, number> = {}

  if (!externalIdColumn && !dbIdColumn) {
    for (const row of rows) {
      unsafeRows.push(row)
      countReason(unsafeReasonCounts, 'missing_key_mapping')
    }
    return {
      keyType,
      keyColumn: null,
      safeRows,
      unsafeRows,
      unsafeReasonCounts,
    }
  }

  for (const row of rows) {
    const externalIdValue = readTrimmedValue(row, externalIdColumn)
    const databaseIdValue = readTrimmedValue(row, dbIdColumn)

    if (externalIdValue.length > 0) {
      safeRows.push(row)
      continue
    }

    if (dbIdColumn) {
      if (databaseIdValue.length === 0) {
        unsafeRows.push(row)
        countReason(
          unsafeReasonCounts,
          externalIdColumn ? 'empty_external_and_database_id' : 'empty_database_id'
        )
        continue
      }

      if (isValidDatabaseId(databaseIdValue)) {
        safeRows.push(row)
        continue
      }

      unsafeRows.push(row)
      countReason(unsafeReasonCounts, 'invalid_database_id')
      continue
    }

    unsafeRows.push(row)
    countReason(unsafeReasonCounts, 'empty_external_id')
  }

  return {
    keyType,
    keyColumn,
    safeRows,
    unsafeRows,
    unsafeReasonCounts,
  }
}

File: ametras_fast_import_addon/models/import_engine/idempotency.py
"""
Row-level idempotency assessment for standalone (RPC) retry decisions.

Port of 16.0's assessTimeoutRetryIdempotency (TypeScript) to Python.

A row is *safe to retry* if it carries a non-empty external ID ('id') or
database ID ('.id') that the import engine will use for upsert — Odoo's
load() will update the existing record rather than creating a duplicate.

Rows without a key column mapping, or with an empty key value, are *unsafe*:
retrying them risks creating duplicate records.

This check is only consulted in standalone (RPC) mode, and only when the
BatchSizeAdapter is already at minimum batch size.  In embedded mode, ORM
savepoints make every batch idempotent at the row level.
"""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from .parser import ParsedRow


@dataclass
class IdempotencyResult:
    safe: list[ParsedRow] = field(default_factory=list)
    unsafe: list[ParsedRow] = field(default_factory=list)
    reason_counts: dict[str, int] = field(default_factory=dict)


def assess_timeout_retry_idempotency(
    rows: list[ParsedRow],
    field_mappings: dict[str, str],
) -> IdempotencyResult:
    """
    Classify rows in *rows* as safe or unsafe to retry after a timeout.

    Args:
        rows: Batch of parsed rows to classify.
        field_mappings: CSV-column → Odoo-field mapping dict
                        (e.g. {'ID': 'id', 'Ext ID': 'id', ...}).

    Returns:
        IdempotencyResult with safe, unsafe lists and per-reason counts.

    Classification:
    - All rows → unsafe with reason 'missing_key_mapping' when neither 'id'
      nor '.id' appears as a mapped target field.
    - Otherwise, per row: if the CSV value for the id or .id column is
      non-empty → safe; else → unsafe with 'empty_external_and_database_id'.
    """
    result = IdempotencyResult()

    # Find CSV column names that map to 'id' and '.id' target fields
    id_col = next((col for col, tgt in field_mappings.items() if tgt == "id"), None)
    dot_id_col = next(
        (col for col, tgt in field_mappings.items() if tgt == ".id"), None
    )

    if id_col is None and dot_id_col is None:
        # No key mapping at all — retrying any row risks a duplicate
        for row in rows:
            result.unsafe.append(row)
        result.reason_counts["missing_key_mapping"] = len(rows)
        return result

    for row in rows:
        id_val = str(row.data.get(id_col, "")).strip() if id_col else ""
        dot_id_val = str(row.data.get(dot_id_col, "")).strip() if dot_id_col else ""

        if id_val or dot_id_val:
            result.safe.append(row)
        else:
            result.unsafe.append(row)
            key = "empty_external_and_database_id"
            result.reason_counts[key] = result.reason_counts.get(key, 0) + 1

    return result
